Source/dom/export.js

```diff
@@ -46,58 +46,146 @@ export const DEFAULT_EXPORT_OPTIONS = {
+  }
+
+  function archiveNativeObject(nativeObject) {
+    const archiver = MSJSONDataArchiver.new()
+    archiver.archiveObjectIDs = true
+    const aPtr = MOPointer.alloc().init()
+    const obj = nativeObject.immutableModelObject
+      ? nativeObject.immutableModelObject()
+      : nativeObject
+    archiver.archivedDataWithRootObject_error(obj, aPtr)
+    if (aPtr.value()) {
+      throw Error('Archive error')
+    }
+    return archiver.archivedData()
+  }
+
+  // Return data if no output directory specified
+  if (!merged.output) {
+    const firstObject = objectsToExport[0]
+    if (exportJSON) {
+      const str = NSString.alloc().initWithData_encoding(
+        archiveNativeObject(firstObject),
+        NSUTF8StringEncoding
+      )
+      return JSON.parse(str)
+    }
+    // Input code for returning image data here...
+    throw new Error('Return output is only support for the json format')
+  }
+  // Save files to directory at options.output
   const exporter = MSSelfContainedHighLevelExporter.alloc().initWithOptions(
     merged
   )
 
   function exportNativeLayers(layers) {
     exporter.exportLayers(layers)
   }
-
   function exportNativePage(page) {
     exporter.exportPage(page)
   }
 
-  if (Array.isArray(object)) {
-    const isArrayOfPages = isWrappedObject(object[0])
-      ? object[0].type === Types.Page
-      : String(object[0].class()) === 'MSPage'
+  // Export all object json
+  if (exportJSON) {
+    const fm = NSFileManager.defaultManager()
+    const directory = NSString.stringWithString(
+      merged.output
+    ).stringByExpandingTildeInPath()
+    const comps = String(directory).split('/')
+    fm.createDirectoryAtPath_withIntermediateDirectories_attributes_error(
+      directory,
+      true,
+      null,
+      null
+    )
 
-    if (isArrayOfPages) {
-      // support an array of pages
-      object.forEach(o => {
-        if (isWrappedObject(o)) {
-          exportNativePage(o.sketchObject)
-        } else {
-          exportNativePage(o)
-        }
-      })
-    } else {
-      // support an array of layers
-      exportNativeLayers(
-        object.map(o => {
-          if (isWrappedObject(o)) {
-            return o.sketchObject
-          }
-          return o
-        })
-      )
+    objectsToExport.forEach(o => {
+      const name =
+        merged['use-id-for-name'] === true || !o.name ? o.objectID() : o.name()
+      const pathComps = comps.slice()
+      pathComps.push(`${name}.json`)
+      const url = NSURL.fileURLWithPath(pathComps.join('/'))
+      const data = archiveNativeObject(o)
+      const writeOptions = merged.overwriting
+        ? 0
+        : NSDataWritingWithoutOverwriting
+      const ptr = MOPointer.new()
+      if (!data.writeToURL_options_error(url, writeOptions, ptr)) {
+        throw new Error(`Error writing json file ${ptr.value()}`)
+      }
+    })
+    // If only JSON stop to bypass the exporters PNG default
+    if (merged.formats.length == 0) {
```

Comment:
formats can be undefined

```suggestion
    if (!(merged.formats || {}).length) {
```